lychee-bin/src/commands/check.rs

```diff
@@ -283,6 +285,19 @@ fn show_progress(
     Ok(())
 }
 
+fn get_failed_urls(stats: &mut ResponseStats) -> Vec<(InputSource, Url)> {
+    stats
+        .fail_map
+        .iter()
+        .flat_map(|(source, set)| set.iter().map(|entry| (source, entry)).collect::<Vec<_>>())
```

Comment:
Thank you, I have removed the `collect` now.
But I'm not really sure how I could combine lines 292-297.
If I rewrite the lines with a for loop I would get something like the following code, but I don't know how I could further simplify it.

```rust
let mut result = Vec::new();
for (source, set) in stats.fail_map.iter() {
    for entry in set {
        let uri = &entry.uri;
        if !(uri.is_data() || uri.is_mail() || uri.is_file()) {
            result.push((source.clone(), entry.uri.as_str().try_into().unwrap()));
        }
    }
}
result
```

Reply:
Okay. Let me try to explain this

We have an iterator of `(InputSource, HashSet<ResponseBody>)`.

L292 transforms from this to a flattened iterator of `(InputSource, ResponseBody)`

L293-296 filters out some `ResponseBody` based on its `uri`. (`ResponseBody` is just `{ Uri, Status }`)

L297 tries to convert `Uri` into `Url`.

It can be simplified as this:

- First, what we need is only `Uri` of `ResponseBody`. So the `flat_map` on L292 can be changed accrodingly.
- Second, we can combine `filter` on L293-296 and `map` on L297 into a single `filter_map` that:
  - If the `Uri` passes the sets of conditionals, try to convert it to a `Url`.
  - Otherwise, simply yield a `None`.

So L292 becomes

```rust
        .flat_map(|(source, set)| set.iter().map(move |ResponseBody {uri, _status}| (source, uri)))
```

L293-297 becomes

```rust
        .filter_map(|(source, uri)| {
            if uri.is_data() || uri.is_mail() || uri.is_file() {
                None
            } else {
                Some((source.clone(), Url::try_from(uri.as_str()).unwrap()))
            }
        })
```

Reply:
I wrote these by hand so there might be some issues or typos.

Reply:
@thomas-zahner, would you say that you have the bandwidth to make these changes? If not we can take over or merge it as is.

Reply:
I saw that you created the new version which is great. I think I will be able to continue on this at the end of this week or next week.

Reply:
Sounds good, no rush. 😉

Reply:
@lebensterben Oh I see, thank you very much for the detailed explanation. I have now updated it.

@mre I have now added the CLI test and synced the branch with master, so it should be ready to be reviewed and merged.